lib/doctest.js

```diff
@@ -507,60 +520,60 @@ const log = results => {
-  }
-
-  const type_ = type == null ? inferType (path) : type;
-  if (type_ == null) {
-    return Promise.reject (new Error (
-      'Cannot infer type from extension'
-    ));
-  }
-
-  const rewriters = {coffee: rewrite$coffee, js: rewrite$js};
-  const source = toModule (
-    rewriters[type_] ({prefix,
-                       openingDelimiter,
-                       closingDelimiter,
-                       sourceType: 'script'})
-                     (fs.readFileSync (path, 'utf8')
-                      .replace (/\r\n?/g, '\n')
-                      .replace (/^#!.*/, '')),
-    module
+const test = options => path => rewrite => evaluate => {
+  const source = (
+    rewrite ({prefix: options.prefix || '',
+              openingDelimiter: options.openingDelimiter,
+              closingDelimiter: options.closingDelimiter})
+            (fs.readFileSync (path, 'utf8')
+             .replace (/\r\n?/g, '\n')
+             .replace (/^#!.*/, ''))
   );
-
-  if (print) {
+  if (options.print) {
     console.log (source.replace (/\n$/, ''));
     return Promise.resolve ([]);
-  } else if (silent) {
-    return Promise.resolve (evaluate (module, source, path));
   } else {
-    console.log (`running doctests in ${path}...`);
-    const results = evaluate (module, source, path);
-    log (results);
-    return Promise.resolve (results);
+    return evaluate (source)
+      .then (results => {
+        if (!options.silent) {
+          console.log (`running doctests in ${path}...`);
+          log (results);
+        }
+        return results;
+      });
   }
 };
 
-module.exports.rewrite$js = rewrite$js;
-
-module.exports.run = run;
-
-module.exports.log = log;
+export default options => async path => {
+  if (!([undefined, 'amd', 'commonjs', 'esm'].includes (options.module))) {
+    throw new Error (`Invalid module \`${options.module}'`);
+  }
+  if (!([undefined, 'coffee', 'js'].includes (options.type))) {
+    throw new Error (`Invalid type \`${options.type}'`);
+  }
+  if (options.module === 'esm') {
+    if (options.type != null) {
+      throw new Error ('Cannot use file type when module is "esm"');
+    }
+    return test (options)
+                (path)
+                (compose (compose (wrapModule))
+                         (rewriteJs ('module')))
+                (evaluateModule (path));
+  } else {
+    const ext = extname (path);
+    if (options.type == null && !(['.coffee', '.js'].includes (ext))) {
+      throw new Error ('Cannot infer type from extension');
+    }
+    return test (options)
+                (path)
+                ((options.type === 'coffee' ||
+                  options.type == null && ext === '.coffee') ?
+                 rewriteCoffee :
+                 compose (compose (toModule (options.module)))
+                         (rewriteJs ('script')))
+                (options.module === 'commonjs' ?
+                 commonjsEval (path) :
+                 functionEval);
+  }
+};
```

Comment:
Aldwin, I greatly reduced the amount of repetition. I hope you like the changes. :)